### PR TITLE
`Programming exercises`: Allow starting practice mode from the last graded version

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

When the student wants to practice after the due date, they should have the option to either start from scratch or keep going from their graded participation if they did not get 100% there.

### Description

The student can now choose to start from the graded participation when starting the practice mode.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 2 Programming Exercises (one with a participation before the duedate one without)

1. Log in to Artemis
2. Participate before the deadline in the one exercise
3. Check that you can start the practice mode after the deadline in both exercises and have the option to use the graded participation in one of them

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [ ] Test 1 (Should have the same UI and behavior as before)

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| start-practice-mode-button.component.ts | 77% | 90% |
| course-exercise.service.ts | 72% | 96% |
| ParticipationService.java | 78% | 92% |

### Screenshots

Bildschirmfoto 2022-10-13 um 11 53 30

### CODE REVIEW

**Review by ge65cer — COMMENTED** [review]
Nice feature! I found an issue while testing:
Submitting a graded submission before the due date, starting practice mode after the due date and then changing the due date to later (e.g., extending the working time) I get an error when clicking the "Resume exercise" button.

Screenshot 2022-10-18 at 17 39 58

```
{"type":"https://www.jhipster.tech/problem/problem-with-message","title":"Internal Server Error","status":500,"detail":"could not execute batch; SQL [update participation set exercise_id=?, individual_due_date=?, initialization_date=?, initialization_state=?, test_run=?, presentation_score=?, student_id=?, team_id=?, branch=?, build_plan_id=?, repository_url=? where id=?]; constraint [participation.UK6skuc03gvp2gw5lyji5lu0v0v]; nested exception is org.hibernate.exception.ConstraintViolationException: could not execute batch","path":"/api/exercises/221/resume-programming-participation/5294","message":"error.http.500"}
```

File: src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java [hunk 1]
```diff
@@ -195,7 +195,37 @@ private StudentParticipation createNewParticipationWithInitializationDate(Exerci
...
+     * @param optionalGradedStudentParticipation the graded participation of that student, if present
+     * @param useGradedParticipation flag if the graded student participation should be used as baseline for the new repository
+     * @return started participation
+     */
+    private StudentParticipation startPracticeMode(ProgrammingExercise exercise, ProgrammingExerciseStudentParticipation participation, boolean setInitializationDate,
+            Optional<StudentParticipation> optionalGradedStudentParticipation, boolean useGradedParticipation) {
+        // Step 1a) create the student repository (based on the template repository or graded participation)
+        if (useGradedParticipation && optionalGradedStudentParticipation.isPresent()
+                && optionalGradedStudentParticipation.get() instanceof ProgrammingExerciseStudentParticipation programmingExerciseStudentParticipation) {
+            participation = copyRepository(exercise, programmingExerciseStudentParticipation.getVcsRepositoryUrl(), participation);
```
[the remark this answers is not in the document]
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/ParticipationService.java:219]: This was already part of the code before hand and we reassign the parameter multiple times in this code. So I'll keep the style here

**Comment by JohannesStoehr** [comment]
> Nice feature! I found an issue while testing: Submitting a graded submission before the due date, starting practice mode after the due date and then changing the due date to later (e.g., extending the working time) I get an error when clicking the "Resume exercise" button.

This is a bigger underlaying problem. There cannot be two participations in the same exercise for a single user with the same initialization state. So the practice participation would have to be set to another state.
Additionally all hidden tests after the due date would have been leaked in your scenario and potential submission policies would also have been circumvented by allowing the practice participations.

Fixing all of this needs another rather big PR that is definitely out of scope for this one.

**Review by stefanwaldhauser — APPROVED** [review]
Works nicely except the issue @ge65cer mentiomed! Nice work

**Review by Santia-go — APPROVED** [review]
Tested on TS1. 
Working as expected with exercises with deadlines and without. Even for exercises with manual assessment. 
Screenshot 2022-10-26 at 14 57 09
Screenshot 2022-10-26 at 14 57 21

**Comment by krusche** [comment]
Before we merge: can you please double check that the chances do not affect the exam mode?

**Comment by JohannesStoehr** [comment]
Thanks for the reviews. One more review making sure that the exam mode is untouched would be great

**Review by Strohgelaender — APPROVED** [review]
Tested on TS1, I was not able to access the practice button and this menu when participating in an exam.

**Pull request merged**